### PR TITLE
Adding Reasons property in JeaSessionConfiguration and JeaRoleCapabilities resources

### PR DESCRIPTION
**Pull request by NicolasBn**
#### Pull Request (PR) description

- Adding Reason class.
- Adding Reasons property in JeaSessionConfiguration and JeaRoleCapabilities resources.
  It's a requirement of [Guest Configuration](https://docs.microsoft.com/en-us/azure/governance/policy/how-to/guest-configuration-create#get-targetresource-requirements)
- Adding pester tests to check Reasons property.

Need to wait the next release of [DscResource.Common](https://github.com/dsccommunity/DscResource.Common). It contains the Compare-DscParameterState function.

Edit : The stringdata reasons in .strings.psd1 files are generic. They could be update in future. 

#### Task list

- [x] Added an entry under the Unreleased section of the change log in the CHANGELOG.md.
      Entry should say what was changed, and how that affects users (if applicable).
- [ ] Resource documentation added/updated in README.md in resource folder.
- [ ] Resource parameter descriptions added/updated in schema.mof
      and comment-based help.
- [ ] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [ ] Examples appropriately added/updated.
- [ ] Unit tests added/updated. See [DSC Resource Testing Guidelines](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md).
- [x] Integration tests added/updated (where possible). See [DSC Resource Testing Guidelines](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md).
- [x] New/changed code adheres to [DSC Resource Style Guidelines](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md) and [Best Practices](https://github.com/PowerShell/DscResources/blob/master/BestPractices.md).

---
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/jeadsc/46)

### CODE REVIEW

File: source/Classes/JeaRoleCapabilities.ps1 [hunk 1]
```diff
@@ -123,10 +127,49 @@ class JeaRoleCapabilities:RoleCapabilitiesUtility
                 }
             }
             $currentState.Ensure = [Ensure]::Present
+
+            # Compare current and desired state to add reasons
+            $valuesToCheck = $this.psobject.properties.name.where({$_ -notin 'Path','Reasons'})
```
Comment (raandree) [inline source/Classes/JeaRoleCapabilities.ps1:132]: Just a minor formatting thing: .Properties.Name.Where

File: source/Classes/JeaRoleCapabilities.ps1 [hunk 2]
```diff
@@ -123,10 +127,49 @@ class JeaRoleCapabilities:RoleCapabilitiesUtility
...
+
+            $compareState = Compare-DscParameterState `
+                -CurrentValues ($currentState | Convert-ObjectToHashtable) `
+                -DesiredValues ($this | Convert-ObjectToHashtable) `
+                -ValuesToCheck $valuesToCheck | Where-Object {$_.InDesiredState -eq $false }
+
+            $currentState.reasons = switch ($compareState)
```
Comment (raandree) [inline source/Classes/JeaRoleCapabilities.ps1:139]: .reason should start with a capital letter.

File: source/Classes/JeaSessionConfiguration.ps1 [hunk 3]
```diff
@@ -285,6 +297,38 @@ class JeaSessionConfiguration:SessionConfigurationUtility
             }
         }
 
+        # Compare current and desired state to add reasons
+        $valuesToCheck = $this.psobject.properties.name.where({$_ -notin 'Name','Reasons'})
```
Comment (raandree) [inline source/Classes/JeaSessionConfiguration.ps1:301]: Just a minor formatting thing: .Properties.Name.Where

File: source/Classes/JeaSessionConfiguration.ps1 [hunk 4]
```diff
@@ -285,6 +297,38 @@ class JeaSessionConfiguration:SessionConfigurationUtility
...
+
+        $compareState = Compare-DscParameterState `
+            -CurrentValues ($currentState | Convert-ObjectToHashtable) `
+            -DesiredValues ($this | Convert-ObjectToHashtable) `
+            -ValuesToCheck $valuesToCheck | Where-Object {$_.InDesiredState -eq $false }
+
+        $currentState.reasons = switch ($compareState)
```
Comment (raandree) [inline source/Classes/JeaSessionConfiguration.ps1:308]: .reason should start with a capital letter.

**Review by raandree — CHANGES_REQUESTED** [review]
Thanks for the code, looks great!

Just found two minor style issues.

**Review by raandree — APPROVED** [review]
Perfect, thanks!

**Comment by NicolasBn** [comment]
Thanks for your review and your compliment !

**Pull request merged**